cpp/src/fil/common.cuh

```diff
@@ -165,6 +166,103 @@ struct predict_params : shmem_size_params {
   int num_blocks;
 };
 
+namespace dispatch {
```

Comment:
Given the complexity of this template dispatch mechanism, I'd clearly prefer it in a separate pull request.